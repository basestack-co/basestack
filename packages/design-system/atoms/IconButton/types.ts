import { SpaceProps, FlexboxProps, PositionProps } from "styled-system";

export type Variant = "primary" | "primaryNeutral" | "secondary" | "neutral";

export interface IconButtonProps
  extends SpaceProps,
    FlexboxProps,
    PositionProps {
  /**
   * Change button variant
   */
  variant?: Variant;
  /**
   * onClick function
   */
  onClick: () => void;
  /**
   * Optional Icon
   */
  icon: string;
}
