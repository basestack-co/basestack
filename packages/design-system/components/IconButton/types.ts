import React from "react";
import { SpaceProps, FlexboxProps, PositionProps } from "styled-system";

export type Size = "xLarge" | "large" | "medium" | "small";

export type Variant = "primary" | "primaryNeutral" | "secondary" | "neutral";

export interface IconButtonProps
  extends SpaceProps,
    FlexboxProps,
    PositionProps {
  /**
   * Change button variant
   */
  variant?: Variant;
  /**
   * onClick function
   */
  onClick: (e: React.MouseEvent<HTMLButtonElement>) => void;
  /**
   * Icon
   */
  icon: string;
  /**
   * Optional size
   */
  size?: Size;
}
