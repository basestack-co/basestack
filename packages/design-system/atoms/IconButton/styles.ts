import styled, { css } from "styled-components";
import { flexbox, space, position, compose } from "styled-system";
import { rem, darken } from "polished";
import { Variant } from "./types";

const sharedStyles = css`
  ${compose(flexbox, space, position)};
  border: none;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  border-radius: 50%;
  transition: background-color 0.1s ease-in-out, color 0.1s ease-in-out,
    border-color 0.1s ease-in-out;
`;

const primaryStyles = css`
  ${sharedStyles};
  background-color: ${({ theme }) => theme.colors.primary};
  color: ${({ theme }) => theme.colors.white};

  &:hover:not(:active) {
    background-color: ${({ theme }) => darken(0.1, theme.colors.primary)};
  }
`;

const secondaryStyles = css`
  ${sharedStyles};
  background-color: ${({ theme }) => theme.colors.gray100};
  color: ${({ theme }) => theme.colors.white};

  &:hover:not(:active) {
    background-color: ${({ theme }) => theme.colors.gray200};
  }
`;

const neutralStyles = css`
  ${sharedStyles};
  background-color: transparent;
  color: ${({ theme }) => theme.colors.black};

  .material-symbols-sharp {
    color: inherit;
  }
  &:hover {
    background-color: ${({ theme }) => theme.colors.gray100};
  }
  &:active {
    background-color: ${({ theme }) => theme.colors.gray200};
  }
`;

const primaryNeutralStyles = css`
  ${sharedStyles};
  background-color: transparent;
  color: ${({ theme }) => theme.colors.black};

  .material-symbols-sharp {
    color: inherit;
  }
  &:hover {
    color: ${({ theme }) => theme.colors.blue400};
    background-color: ${({ theme }) => theme.colors.blue50};
  }
  &:active {
    background-color: ${({ theme }) => theme.colors.blue100};
  }
`;

const handleButtonVariant = (variant?: Variant) => {
  switch (variant) {
    case "primary":
      return primaryStyles;
    case "primaryNeutral":
      return primaryNeutralStyles;
    case "secondary":
      return secondaryStyles;
    default:
    case "neutral":
      return neutralStyles;
  }
};

export const StyledButton = styled.button<{ variant: Variant }>`
  ${({ variant }) => handleButtonVariant(variant)};
  height: ${rem("36px")};
  width: ${rem("36px")};
`;
