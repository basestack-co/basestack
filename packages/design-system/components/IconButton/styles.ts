import { rem } from "polished";
import type { Ref } from "react";
import styled, { css } from "styled-components";
import { compose, flexbox, position, space } from "styled-system";
import type { Variant } from "./types";

const sharedStyles = css`
  ${compose(flexbox, space, position)};
  border: none;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  border-radius: 50%;
  transition:
    background-color 0.1s ease-in-out,
    color 0.1s ease-in-out,
    border-color 0.1s ease-in-out;
`;

const primaryStyles = css`
  ${sharedStyles};
  background-color: ${({ theme }) => theme.iconButton.primary.backgroundColor};
  color: ${({ theme }) => theme.iconButton.primary.color};

  &:hover:not(:active):not(:disabled) {
    background-color: ${({ theme }) =>
      theme.iconButton.primary.hover.backgroundColor};
  }
`;

const secondaryStyles = css`
  ${sharedStyles};
  background-color: ${({ theme }) =>
    theme.iconButton.secondary.backgroundColor};
  color: ${({ theme }) => theme.iconButton.secondary.color};

  &:hover:not(:active):not(:disabled) {
    background-color: ${({ theme }) =>
      theme.iconButton.secondary.hover.backgroundColor};
  }
`;

const secondaryDarkStyles = css`
  ${sharedStyles};
  background-color: ${({ theme }) =>
    theme.iconButton.secondaryDark.backgroundColor};
  color: ${({ theme }) => theme.iconButton.secondaryDark.color};

  &:hover:not(:active):not(:disabled) {
    background-color: ${({ theme }) =>
      theme.iconButton.secondaryDark.hover.backgroundColor};
  }
`;

const neutralStyles = css`
  ${sharedStyles};
  background-color: ${({ theme }) => theme.iconButton.neutral.backgroundColor};
  color: ${({ theme }) => theme.iconButton.neutral.color};

  .material-symbols-sharp,
  .material-symbols-rounded {
    color: inherit;
  }

  &:hover:not(:disabled) {
    background-color: ${({ theme }) =>
      theme.iconButton.neutral.hover.backgroundColor};
  }

  &:active:not(:disabled) {
    background-color: ${({ theme }) =>
      theme.iconButton.neutral.active.backgroundColor};
  }
`;

const primaryNeutralStyles = css`
  ${sharedStyles};
  background-color: ${({ theme }) =>
    theme.iconButton.primaryNeutral.backgroundColor};
  color: ${({ theme }) => theme.iconButton.primaryNeutral.color};

  .material-symbols-sharp,
  .material-symbols-rounded {
    color: inherit;
  }

  &:hover {
    color: ${({ theme }) => theme.iconButton.primaryNeutral.hover.color};
    background-color: ${({ theme }) =>
      theme.iconButton.primaryNeutral.hover.backgroundColor};
  }

  &:active {
    background-color: ${({ theme }) =>
      theme.iconButton.primaryNeutral.active.backgroundColor};
  }
`;

const handleButtonVariant = (variant?: Variant) => {
  switch (variant) {
    case "primary":
      return primaryStyles;
    case "primaryNeutral":
      return primaryNeutralStyles;
    case "secondary":
      return secondaryStyles;
    case "secondaryDark":
      return secondaryDarkStyles;
    default:
      return neutralStyles;
  }
};

export const StyledButton = styled.button.withConfig({
  shouldForwardProp: (prop) =>
    !["iconSize", "zIndex", "flexShrink", "ref"].includes(prop),
})<{
  variant: Variant;
  iconSize: string;
  ref?: Ref<HTMLDivElement>;
}>`
  ${({ variant }) => handleButtonVariant(variant)};
  height: ${({ iconSize }) => rem(iconSize)};
  width: ${({ iconSize }) => rem(iconSize)};

  &:disabled {
    opacity: 0.5;
    cursor: initial;
  }
`;
