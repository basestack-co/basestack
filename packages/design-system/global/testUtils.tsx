import { type RenderOptions, render } from "@testing-library/react";
import type React from "react";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "../theme/lightTheme";

interface Props {
  children: React.ReactNode;
}

const WithThemeProvider = ({ children }: Props) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithThemeProvider, ...options });

export { renderWithTheme };
