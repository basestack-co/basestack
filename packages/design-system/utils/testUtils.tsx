import defaultTheme from "@basestack/design-system/theme/lightTheme";
import isPropValid from "@emotion/is-prop-valid";
import { RenderResult, render } from "@testing-library/react";
import React from "react";
import { StyleSheetManager, ThemeProvider } from "styled-components";

const Provider = ({
  children,
  theme = defaultTheme,
}: {
  children: React.ReactNode;
  theme: any;
}) => (
  <ThemeProvider theme={theme}>
    <StyleSheetManager shouldForwardProp={isPropValid}>
      {children}
    </StyleSheetManager>
  </ThemeProvider>
);

export const renderWithTheme = (
  ui: React.ReactElement,
  options?: any,
): RenderResult => render(ui, { wrapper: Provider, ...options });
