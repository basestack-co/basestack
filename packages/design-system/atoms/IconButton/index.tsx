import { forwardRef, memo } from "react";
import { useTheme } from "styled-components";
import { IconButtonProps } from "./types";
import Icon, { Size as IconSize } from "../Icon";
import { StyledButton } from "./styles";

const IconButton = forwardRef<HTMLButtonElement, IconButtonProps>(
  ({ variant = "neutral", onClick, icon, size, ...props }, ref) => {
    const theme = useTheme();

    const handleIconColor = () => {
      switch (variant) {
        case "primary":
          return theme.colors.white;
        case "primaryNeutral":
          return theme.colors.blue400;
        case "secondary":
          return theme.colors.black;
        default:
        case "neutral":
          return theme.colors.black;
      }
    };

    const handleButtonSize = () => {
      switch (size) {
        case "xLarge":
          return {
            bg: "56px",
            icon: "xLarge", // icon 48
          };
        case "large":
          return {
            bg: "40px",
            icon: "large", // icon 32
          };
        default:
        case "medium":
          return {
            bg: "32px",
            icon: "medium", // icon 24
          };
        case "small":
          return {
            bg: "24px",
            icon: "small", // icon 18
          };
      }
    };

    const iconColor = handleIconColor();
    const buttonSize = handleButtonSize();

    return (
      <StyledButton
        ref={ref}
        onClick={onClick}
        variant={variant}
        iconSize={buttonSize.bg}
        {...props}
      >
        <Icon
          icon={icon}
          size={buttonSize.icon as IconSize}
          color={iconColor}
        />
      </StyledButton>
    );
  },
);

IconButton.displayName = "IconButton";

export default memo(IconButton);
