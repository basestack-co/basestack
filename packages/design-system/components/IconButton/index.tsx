import { memo } from "react";
import { useTheme } from "styled-components";
import Icon, { type Size as IconSize } from "../Icon";
import { StyledButton } from "./styles";
import type { IconButtonProps, Variant as IconButtonVariant } from "./types";

const IconButton = ({
  ref,
  variant = "neutral",
  onClick,
  icon,
  size,
  isDisabled,
  ...props
}: IconButtonProps) => {
  const theme = useTheme();

  const handleIconColor = () => {
    switch (variant) {
      case "primary":
        return theme.iconButton.primary.color;
      case "primaryNeutral":
        return theme.iconButton.primaryNeutral.color;
      case "secondary":
        return theme.iconButton.secondary.color;
      case "secondaryDark":
        return theme.iconButton.secondary.color;
      default:
        return theme.iconButton.neutral.color;
    }
  };

  const handleButtonSize = () => {
    switch (size) {
      case "xLarge":
        return {
          bg: "56px",
          icon: "xLarge", // icon 48
        };
      case "large":
        return {
          bg: "40px",
          icon: "large", // icon 32
        };
      case "mediumLarge":
        return {
          bg: "36px",
          icon: "medium", // icon 24
        };
      case "small":
        return {
          bg: "24px",
          icon: "small", // icon 18
        };
      default:
        return {
          bg: "32px",
          icon: "medium", // icon 24
        };
    }
  };

  const iconColor = handleIconColor();
  const buttonSize = handleButtonSize();

  return (
    <StyledButton
      ref={ref}
      onClick={onClick}
      variant={variant}
      iconSize={buttonSize.bg}
      disabled={isDisabled}
      {...props}
    >
      <Icon icon={icon} size={buttonSize.icon as IconSize} color={iconColor} />
    </StyledButton>
  );
};

IconButton.displayName = "IconButton";

export type { IconButtonProps, IconButtonVariant };

export default memo(IconButton);
