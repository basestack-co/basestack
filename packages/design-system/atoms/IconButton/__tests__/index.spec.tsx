import { cleanup } from "@testing-library/react";
import { renderWithTheme } from "../../../utils/testUtils";
import IconButton from "..";

describe("IconButton Atom tests", () => {
  afterEach(cleanup);

  test("should render IconButton correctly", () => {
    const { asFragment } = renderWithTheme(
      <IconButton onClick={jest.fn()} icon="help" />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("should render primary IconButton", () => {
    const { asFragment } = renderWithTheme(
      <IconButton onClick={jest.fn()} icon="help" variant="primary" />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("should render primaryNeutral IconButton", () => {
    const { asFragment } = renderWithTheme(
      <IconButton onClick={jest.fn()} icon="help" variant="primaryNeutral" />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("should render secondary IconButton", () => {
    const { asFragment } = renderWithTheme(
      <IconButton onClick={jest.fn()} icon="help" variant="secondary" />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("should render IconButton with different size", () => {
    const { asFragment } = renderWithTheme(
      <IconButton
        onClick={jest.fn()}
        icon="help"
        variant="secondary"
        size="large"
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
