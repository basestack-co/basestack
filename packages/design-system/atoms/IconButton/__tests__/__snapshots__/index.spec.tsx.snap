// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton Atom tests should render IconButton correctly 1`] = `
<DocumentFragment>
  .c1 {
  height: 1.5rem;
  width: 1.5rem;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2 {
  color: #000000;
  font-size: 1.5rem;
}

.c0 {
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  border-radius: 50%;
  -webkit-transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  background-color: transparent;
  color: #000000;
  height: 2.25rem;
  width: 2.25rem;
}

.c0 .material-symbols-sharp {
  color: inherit;
}

.c0:hover {
  background-color: #EEEEEE;
}

.c0:active {
  background-color: #E2E2E2;
}

<button
    class="c0"
  >
    <div
      class="c1"
      data-testid="icon-container"
      size="24"
    >
      <span
        class="c2 material-symbols-sharp"
        color="#000000"
        data-testid="icon"
        size="24"
      >
        help
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`IconButton Atom tests should render primary IconButton 1`] = `
<DocumentFragment>
  .c1 {
  height: 1.5rem;
  width: 1.5rem;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2 {
  color: #FFFFFF;
  font-size: 1.5rem;
}

.c0 {
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  border-radius: 50%;
  -webkit-transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  background-color: #276EF1;
  color: #FFFFFF;
  height: 2.25rem;
  width: 2.25rem;
}

.c0:hover:not(:active) {
  background-color: #0e55d7;
}

<button
    class="c0"
  >
    <div
      class="c1"
      data-testid="icon-container"
      size="24"
    >
      <span
        class="c2 material-symbols-sharp"
        color="#FFFFFF"
        data-testid="icon"
        size="24"
      >
        help
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`IconButton Atom tests should render primaryNeutral IconButton 1`] = `
<DocumentFragment>
  .c1 {
  height: 1.5rem;
  width: 1.5rem;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2 {
  color: #276EF1;
  font-size: 1.5rem;
}

.c0 {
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  border-radius: 50%;
  -webkit-transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  background-color: transparent;
  color: #000000;
  height: 2.25rem;
  width: 2.25rem;
}

.c0 .material-symbols-sharp {
  color: inherit;
}

.c0:hover {
  color: #276EF1;
  background-color: #EFF3FE;
}

.c0:active {
  background-color: #D4E2FC;
}

<button
    class="c0"
  >
    <div
      class="c1"
      data-testid="icon-container"
      size="24"
    >
      <span
        class="c2 material-symbols-sharp"
        color="#276EF1"
        data-testid="icon"
        size="24"
      >
        help
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`IconButton Atom tests should render secondary IconButton 1`] = `
<DocumentFragment>
  .c1 {
  height: 1.5rem;
  width: 1.5rem;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2 {
  color: #000000;
  font-size: 1.5rem;
}

.c0 {
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  border-radius: 50%;
  -webkit-transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out, border-color 0.1s ease-in-out;
  background-color: #EEEEEE;
  color: #FFFFFF;
  height: 2.25rem;
  width: 2.25rem;
}

.c0:hover:not(:active) {
  background-color: #E2E2E2;
}

<button
    class="c0"
  >
    <div
      class="c1"
      data-testid="icon-container"
      size="24"
    >
      <span
        class="c2 material-symbols-sharp"
        color="#000000"
        data-testid="icon"
        size="24"
      >
        help
      </span>
    </div>
  </button>
</DocumentFragment>
`;
